client/asset/dcr/dcr.go

```diff
@@ -3078,6 +3091,296 @@ func (dcr *ExchangeWallet) EstimateRegistrationTxFee(feeRate uint64) uint64 {
 	return (dexdcr.MsgTxOverhead + dexdcr.P2PKHOutputSize*2 + inputCount*dexdcr.P2PKHInputSize) * feeRate
 }
 
+// MakeBondTx creates a time-locked fidelity bond transaction. The V0
+// transaction has two required outputs:
+//
+// Output 0 is a the time-locked bond output of type P2SH with the provided
+// value. The redeem script looks similar to the refund path of an atomic swap
+// script:
+//
+//	<locktime[4]> OP_CHECKLOCKTIMEVERIFY OP_DROP <pubkey[33]> OP_CHECKSIG
+//
+// The pubkey referenced by the script comes from the wallet, but in general, it
+// need not be controlled by the wallet.
+//
+// Output 1 is a DEX Account commitment. This is an OP_RETURN output that
+// references the provided account ID.
+//
+//	OP_RETURN <account ID>
+//
+// Having the account ID in the raw allows the txn alone to identify the account
+// without the bond output's redeem script.
+//
+// Output 2 is change, if any.
+//
+// The bond output's redeem script, which is needed to spend the bond output and
+// which the server will need to verify that the bond is redeemable by the
+// account owner, is returned. The bond output pays to a pubkey script for a
+// wallet change address, and the corresponding private key is used to sign a
+// message, sha256(account ID). This is the Bond.BondAcctSig field, which the
+// server should verify using the account ID and pubkey recovered from the
+// signature via ecdsa.RecoverCompact(sig, msg). Bond.RedeemTx is a backup
+// transaction that spends the bond output after lockTime passes, paying to an
+// address for the current underlying wallet; the bond private key (BondPrivKey)
+// should normally be used to author a new transaction paying to a new address
+// instead.
+//
+// NOTE: Alternatively, the caller could provide the pubKey, where the private
+// key is managed by the caller instead of the wallet. The caller would also
+// create the BondAcctSig using a signature algorithm and message (e.g. acct ID
+// hash) of their choice.
+func (dcr *ExchangeWallet) MakeBondTx(ver uint16, amt uint64, lockTime time.Time, acctID []byte) (*asset.Bond, error) {
+	if ver != 0 {
+		return nil, errors.New("only version 0 bonds supported")
+	}
+	if until := time.Until(lockTime); until >= 365*12*time.Hour /* ~6 months */ {
+		return nil, fmt.Errorf("that lock time is nuts: %v", lockTime)
+	} else if until < 0 {
+		return nil, fmt.Errorf("that lock time is already passed: %v", lockTime)
+	}
+
+	// Get an address to which the time-locked fidelity bond should pay. This
+	// address' keys are also used to sign a message (sha256 of account ID) for
+	// the server to verify the acct owns this bond. NOTE: Instead of a wallet
+	// key pair for the bond script, we could pay to the account pubkey, but
+	// that would be poor security.
+	addr, err := dcr.wallet.InternalAddress(dcr.ctx, dcr.config().primaryAcct)
```

Comment:
> It is visible to others outside of the wallet.

What is meant by "visible" here?  Change addresses are just as visible as external addresses when in published transactions.

Reply:
> What is meant by "visible" here?

You are showing the pubkey to the dex, I guess.

I can't say I completely understand, but it seems like internal is just for change addresses. So, the wallet doesn't really need to keep as close track of them. You can also leave it... I dunno if it matters really. Cryptographically there's no difference.

Reply:
In BIP44, the language "visible" I think is a poor wording for sharing this address with other parties so that they can pay to them, vs ones that you intend to use only your self.  In this case, both are like change.

I certainly want to do "the right thing" but I'm having trouble seeing why it matters.

The only consideration is *account* discovery on wallet re-seed, but the address used in a non-standard script isn't going to aid account discovery regardless of which internal/external branch it is from.  In fact, if we use external addresses in these non-standard redeem scripts that the wallet won't find (it only likely recognizes keys used in P2(W)PKH-nested-in-P2SH addresses as per bip49), it might actually be a problem if there's not an external address used in a normal output e.g. initial deposit.

Reply:
Moot point now that the key is from the caller (dex client HD keychain).